solutions/epi/lab3/kern/env.c

```diff
@@ -323,11 +346,42 @@ load_icode(struct Env *e, uint8_t *binary, size_t size)
 	//  What?  (See env_run() and env_pop_tf() below.)
 
 	// LAB 3: Your code here.
+	struct Proghdr *ph, *eph;
+
+	struct Elf *hdr = (struct Elf *)binary;
+
+	// is this a valid ELF?
+	if (hdr->e_magic != ELF_MAGIC)
+	    panic("No elf magic!");
+	cprintf("elf magic!\n");
+
+	// load each program segment (ignores ph flags)
+	// uint32_t cr3 = rcr3();
+	lcr3(PADDR(e->env_pgdir));
+
+	ph = (struct Proghdr *)((uint8_t *)hdr + hdr->e_phoff);
+	eph = ph + hdr->e_phnum;
+
+
+	for (; ph < eph; ph++) {
+	    if (ph->p_type != ELF_PROG_LOAD) continue;
+
+	    region_alloc(e, (void*)ph->p_va, ph->p_memsz);
+	    memcpy((void*)ph->p_va, binary + ph->p_offset, ph->p_filesz);
+	    memset((void *)(ph->p_va + ph->p_filesz), 0, ph->p_memsz - ph->p_filesz);
+	}
+	lcr3(PADDR(kern_pgdir));
+	cprintf("copied\n");
```

Comment:
where it should be printed?

Reply:
It's a debug output marking end of file copying. I think it is where it should be.